syntax = "proto3";
package Simulation;

// ping

message CtrlPingRequest {
  string input = 1;
}

message CtrlPingResponse {
	string output = 1;
}

// list model instances

message ListInstancesRequest {
}

message ListInstancesResponse {
  repeated string instances = 1;
}

// delete model instances

message DeleteInstancesRequest {
  repeated string instances = 1;
}

message DeleteInstancesResponse {
}

// simulation

message StartSimulationRequest {
  double balance = 2;
  repeated string tickers = 3;
  string start_date = 4;
  string end_date = 5;
  repeated string models = 6;
  float transaction_fee = 7;
  int32 number_executions = 8;
}

message ModelResults {
  string model = 1;
  double nominal_profit = 2;
  double percentage_profit = 3;
  double yearly_percentage_profit = 4;
  double operating_time_percentage = 5;
}

message TickerResults {
  string ticker = 1;
  double percentage_profit = 3;
}

message SimulationResult {
  int32 simulation_id = 1;
  repeated ModelResults model_results = 2;
  repeated TickerResults ticker_results = 3;
}

message TickerList{
  repeated string available_tickers = 1;
  repeated string not_available_tickers= 2;
}

message TimeInterval {
  string start_date = 1;
  string end_date = 2;
}

message StartSimulationResponse {
  oneof response {
    SimulationResult simulation_result = 1;
    TickerList tickers = 2;
    TimeInterval interval = 3;
  }
  enum Status {
    OK = 0;
    TICKERS_NOT_AVAILABLE = 1;
    INTERVAL_NOT_AVAILABLE = 2;
    MODEL_INSTANCES_NOT_AVAILABLE = 3;
    NOK = 4;
  }
  Status status = 4;
}

// simulation graphs

message DataPoint {
  int32 time = 1;
  double capital = 2;
}

message SimulationGraphRequest{
  int32 simulation_id = 1;
}

message SimulationGraphResponse{
  string name = 1;
  repeated DataPoint data_points = 2;
}

// simulation logs

message Log {
  enum Action {
    BUY = 0;
    SELL = 1;
  }
  Action action = 1;
  float amount = 2;
  string ticker = 3;
  float price_per_share = 4;
  string date = 5;
}

message SimulationLogsRequest{
  int32 id = 1;
}

message SimulationLogsResponse{
  string model = 1;
  int32 no_execution = 2;
  repeated Log logs = 3;
}

// simulation termination

message CloseSimulationRequest{
  int32 simulation_id = 1;
}

message CloseSimulationResponse{
  enum Status {
    OK = 0;
    NOK = 1;
  }
  Status status = 1;
}

// services

service Simulation {
  rpc ctrl_ping(CtrlPingRequest) returns (CtrlPingResponse);
  rpc list_model_instances(ListInstancesRequest) returns (ListInstancesResponse);
  rpc delete_model_instances(DeleteInstancesRequest) returns (DeleteInstancesResponse);
  rpc start_simulation(StartSimulationRequest) returns (StartSimulationResponse);
  rpc simulation_graph(SimulationGraphRequest) returns (stream SimulationGraphResponse);
  rpc simulation_logs(SimulationLogsRequest) returns (stream SimulationLogsResponse);
  rpc close_simulation(CloseSimulationRequest) returns (CloseSimulationResponse);
}
